pkg/ipam-node/allocator/range.go

```diff
@@ -0,0 +1,181 @@
+	Gateway    net.IP      `json:"gateway,omitempty"`
+}
+
+// Canonicalize takes a given range and ensures that all information is consistent,
+// filling out Start, End, and Gateway with sane values if missing
+func (r *Range) Canonicalize() error {
+	if err := CanonicalizeIP(&r.Subnet.IP); err != nil {
+		return err
+	}
+
+	// Can't create an allocator for a network with no addresses, eg
+	// a /32 or /31
+	ones, masklen := r.Subnet.Mask.Size()
+	if ones > masklen-2 {
+		return fmt.Errorf("network %s too small to allocate from", (*net.IPNet)(&r.Subnet).String())
+	}
+
+	if len(r.Subnet.IP) != len(r.Subnet.Mask) {
+		return fmt.Errorf("IPNet IP and Mask version mismatch")
+	}
+
+	// Ensure Subnet IP is the network address, not some other address
+	networkIP := r.Subnet.IP.Mask(r.Subnet.Mask)
+	if !r.Subnet.IP.Equal(networkIP) {
+		return fmt.Errorf("network has host bits set. "+
+			"For a subnet mask of length %d the network address is %s", ones, networkIP.String())
+	}
+
+	// If the gateway is nil, claim .1
+	if r.Gateway == nil {
+		r.Gateway = ip.NextIP(r.Subnet.IP)
+		if r.Gateway == nil {
+			return fmt.Errorf("computed Gateway for the subnet is not a valid IP")
+		}
+	} else {
+		if err := CanonicalizeIP(&r.Gateway); err != nil {
+			return err
+		}
+	}
+
+	// RangeStart: If specified, make sure it's sane (inside the subnet),
+	// otherwise use the first free IP (i.e. .1) - this will conflict with the
+	// gateway but we skip it in the iterator
+	if r.RangeStart != nil {
+		if err := CanonicalizeIP(&r.RangeStart); err != nil {
+			return err
+		}
+
+		if !r.Contains(r.RangeStart) {
+			return fmt.Errorf("RangeStart %s not in network %s", r.RangeStart.String(), (*net.IPNet)(&r.Subnet).String())
+		}
+	} else {
+		r.RangeStart = ip.NextIP(r.Subnet.IP)
+		if r.RangeStart == nil {
+			return fmt.Errorf("computed RangeStart is not a valid IP")
+		}
+	}
+
+	// RangeEnd: If specified, verify sanity. Otherwise, add a sensible default
+	// (e.g. for a /24: .254 if IPv4, ::255 if IPv6)
+	if r.RangeEnd != nil {
+		if err := CanonicalizeIP(&r.RangeEnd); err != nil {
+			return err
+		}
+
+		if !r.Contains(r.RangeEnd) {
+			return fmt.Errorf("RangeEnd %s not in network %s", r.RangeEnd.String(), (*net.IPNet)(&r.Subnet).String())
+		}
+	} else {
+		r.RangeEnd = lastIP(r.Subnet)
+	}
+
+	return nil
+}
+
+// Contains checks if a given ip is a valid, allocatable address in a given Range
+func (r *Range) Contains(addr net.IP) bool {
+	if err := CanonicalizeIP(&addr); err != nil {
+		return false
+	}
+
+	subnet := (net.IPNet)(r.Subnet)
+
+	// Not the same address family
+	if len(addr) != len(r.Subnet.IP) {
+		return false
+	}
+
+	// Not in network
+	if !subnet.Contains(addr) {
+		return false
+	}
+
+	// We ignore nils here so we can use this function as we initialize the range.
+	if r.RangeStart != nil {
+		// Before the range start
+		if ip.Cmp(addr, r.RangeStart) < 0 {
+			return false
+		}
+	}
+
+	if r.RangeEnd != nil {
```

Comment:
when can we have RangeEnd  as nil ?

Reply:
I didn't find such scenario in our code. I removed this check.